File: radar-commons-android/src/test/java/org/radarcns/android/auth/AppAuthStateTest.java
package org.radarcns.android.auth;

import android.content.Context;
import android.os.Bundle;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.radarcns.android.auth.LoginManager.AUTH_TYPE_BEARER;
import static org.radarcns.android.auth.portal.ManagementPortalClient.MP_REFRESH_TOKEN_PROPERTY;

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public class AppAuthStateTest {
    private AppAuthState state;
    private ArrayList<SourceMetadata> sources;

    @Before
    public void setUp() {
        sources = new ArrayList<>();
        SourceMetadata source = new SourceMetadata(1, "radar", "test", "1.0", true);
        source.setExpectedSourceName("something");
        sources.add(source);

        state = new AppAuthState.Builder()
                .token("abcd")
                .tokenType(AUTH_TYPE_BEARER)
                .projectId("p")
                .userId("u")
                .attribute(MP_REFRESH_TOKEN_PROPERTY, "efgh")
                .sourceMetadata(sources)
                .header("Authorization", "Bearer abcd")
                .expiration(System.currentTimeMillis() + 10_000L)
                .build();

        testProperties(state);
    }

    @Test
    public void addToPreferences() {
        Context context = RuntimeEnvironment.application.getApplicationContext();
        state.addToPreferences(context);

        AppAuthState readState = AppAuthState.Builder.from(context).build();
        testProperties(readState);
    }

    private void testProperties(AppAuthState state) {
        testProperties(state, "efgh");
    }

    private void testProperties(AppAuthState state, String refreshToken) {
        assertEquals("abcd", state.getToken());
        assertEquals(refreshToken, state.getAttribute(MP_REFRESH_TOKEN_PROPERTY));
        assertEquals("p", state.getProjectId());
        assertEquals("u", state.getUserId());
        assertTrue(state.isValidFor(9, TimeUnit.SECONDS));
        assertFalse(state.isValidFor(11, TimeUnit.SECONDS));
        assertEquals(AUTH_TYPE_BEARER, state.getTokenType());
        assertEquals("Bearer abcd", state.getHeaders().get(0).getValue());
        assertEquals(sources, state.getSourceMetadata());
    }

    @Test
    public void addToBundle() {
        Bundle bundle = new Bundle();
        state.addToBundle(bundle);

        AppAuthState readState = AppAuthState.Builder.from(bundle).build();
        testProperties(readState);
    }

    @Test
    public void newBuilder() {
        AppAuthState builtState = state.newBuilder()
                .attribute(MP_REFRESH_TOKEN_PROPERTY, "else")
                .build();

        testProperties(builtState, "else");
        testProperties(state);
    }
}

File: radar-commons-android/src/test/java/org/radarcns/android/auth/AppAuthStateTest.kt
package org.radarcns.android.auth

import android.os.Bundle
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.radarcns.android.auth.LoginManager.AUTH_TYPE_BEARER
import org.radarcns.android.auth.portal.ManagementPortalClient.Companion.MP_REFRESH_TOKEN_PROPERTY
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import java.util.*
import java.util.concurrent.TimeUnit

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
class AppAuthStateTest {
    private lateinit var state: AppAuthState
    private lateinit var sources: ArrayList<SourceMetadata>

    @Before
    fun setUp() {
        sources = ArrayList()
        SourceMetadata(1, "radar", "test", "1.0", true)
                .apply { expectedSourceName = "something" }
                .also { sources.add(it) }

        state = AppAuthState {
            token = "abcd"
            tokenType = AUTH_TYPE_BEARER
            projectId = "p"
            userId = "u"
            attributes[MP_REFRESH_TOKEN_PROPERTY] = "efgh"
            sourceMetadata += sources
            addHeader("Authorization", "Bearer abcd")
            expiration = System.currentTimeMillis() + 10_000L
            isPrivacyPolicyAccepted = true
        }

        testProperties(state)
    }

    @Test
    fun addToPreferences() {
        val readState = RuntimeEnvironment.application.applicationContext.let { context ->
            state.addToPreferences(context)
            AppAuthState.from(context)
        }

        testProperties(readState)
    }

    private fun testProperties(state: AppAuthState, refreshToken: String = "efgh") {
        assertEquals("abcd", state.token)
        assertEquals(refreshToken, state.getAttribute(MP_REFRESH_TOKEN_PROPERTY))
        assertEquals("p", state.projectId)
        assertEquals("u", state.userId)
        assertTrue(state.isValidFor(9, TimeUnit.SECONDS))
        assertFalse(state.isValidFor(11, TimeUnit.SECONDS))
        assertEquals(AUTH_TYPE_BEARER.toLong(), state.tokenType.toLong())
        assertEquals("Bearer abcd", state.headers[0].value)
        assertEquals(sources, state.sourceMetadata)
    }

    @Test
    fun addToBundle() {
        val readState = Bundle().let {
            state.addToBundle(it)
            AppAuthState.from(it)
        }
        testProperties(readState)
    }

    @Test
    fun newBuilder() {
        val builtState = state.alter {
            attributes[MP_REFRESH_TOKEN_PROPERTY] = "else"
        }

        testProperties(builtState, "else")
        testProperties(state)
    }
}
